irr/src/CGLTFMeshFileLoader.cpp

```diff
@@ -0,0 +1,610 @@
+	const core::matrix4 transform = loadTransform(node.transform);
+	joint->LocalMatrix = transform;
+	joint->GlobalMatrix = parent ? parent->GlobalMatrix * joint->LocalMatrix : joint->LocalMatrix;
+	if (node.name.has_value()) {
+		joint->Name = node.name->c_str();
+	}
+	if (node.mesh.has_value()) {
+		loadMesh(*node.mesh, joint);
+	}
+	if (node.children.has_value()) {
+		for (const auto &child : *node.children) {
+			loadNode(child, joint);
+		}
+	}
+}
+
+void CGLTFMeshFileLoader::MeshExtractor::loadNodes() const
+{
+	std::vector<bool> isChild(m_gltf_model.nodes->size());
+	for (const auto &node : *m_gltf_model.nodes) {
+		if (!node.children.has_value())
+			continue;
+		for (const auto &child : *node.children) {
+			isChild[child] = true;
+		}
+	}
+	// Load all nodes that aren't children.
+	// Children will be loaded by their parent nodes.
+	for (std::size_t i = 0; i < m_gltf_model.nodes->size(); ++i) {
+		if (!isChild[i]) {
+			loadNode(i, nullptr);
+		}
+	}
+}
+
+/**
+ * Extracts GLTF mesh indices into the irrlicht model.
+*/
+std::optional<std::vector<u16>> CGLTFMeshFileLoader::MeshExtractor::getIndices(
+		const std::size_t meshIdx,
+		const std::size_t primitiveIdx) const
+{
+	const auto accessorIdx = getIndicesAccessorIdx(meshIdx, primitiveIdx);
+	if (!accessorIdx.has_value())
+		return std::nullopt; // non-indexed geometry
+	const auto &accessor = m_gltf_model.accessors->at(accessorIdx.value());
+	
+	const auto& buf = getBuffer(accessorIdx.value());
+
+	std::vector<u16> indices{};
+	const auto count = getElemCount(accessorIdx.value());
+	for (std::size_t i = 0; i < count; ++i) {
+		std::size_t elemIdx = count - i - 1; // reverse index order
+		u16 index;
+		// Note: glTF forbids the max value for each component type.
+		switch (accessor.componentType) {
+			case tiniergltf::Accessor::ComponentType::UNSIGNED_BYTE: {
+				index = readPrimitive<u8>(BufferOffset(buf, elemIdx * sizeof(u8)));
+				if (index == std::numeric_limits<u8>::max())
+					throw std::runtime_error("invalid index");
+				break;
+			}
+			case tiniergltf::Accessor::ComponentType::UNSIGNED_SHORT: {
+				index = readPrimitive<u16>(BufferOffset(buf, elemIdx * sizeof(u16)));
+				if (index == std::numeric_limits<u16>::max())
+					throw std::runtime_error("invalid index");
+				break;
+			}
+			case tiniergltf::Accessor::ComponentType::UNSIGNED_INT: {
+				u32 indexWide = readPrimitive<u32>(BufferOffset(buf, elemIdx * sizeof(u32)));
+				// Use >= here for consistency.
+				if (indexWide >= std::numeric_limits<u16>::max())
+					throw std::runtime_error("index too large (>= 65536)");
+				index = static_cast<u16>(indexWide);
+				break;
+			}
+			default:
+				throw std::runtime_error("invalid index component type");
+		}
+		indices.push_back(index);
+	}
+
+	return indices;
+}
+
+/**
+ * Create a vector of video::S3DVertex (model data) from a mesh & primitive index.
+*/
+std::optional<std::vector<video::S3DVertex>> CGLTFMeshFileLoader::MeshExtractor::getVertices(
```

Comment:
inconsistent usage of `vertex_t`

Reply:
got rid of `vertex_t` in https://github.com/minetest/minetest/pull/14557/commits/eea01c3957308c7bc1d0a8fe03dd2324447d778d